plugins/optimization/src/main/java/org/qbicc/plugin/opt/ea/EscapeAnalysisIntraMethodBuilder.java

```diff
@@ -154,14 +159,24 @@ public void startMethod(List<ParameterValue> arguments) {
     public BasicBlock return_(Value value) {
         final BasicBlock result = super.return_(value);
 
-        // Skip primitive values truncated, they are not objects
-        if (!(value instanceof Truncate)) {
+        if (value instanceof Call call && !isPrimitive(call.getType())) {
```

Comment:
Other types of calls that can return are `Invoke` and `CallNoSideEffects`; however `CallNoSideEffects` is not allowed to modify memory already so you can exclude that one. But you do need to consider `Invoke`.

Reply:
Hmmm, is that really possible? `Invoke` is not of type `Value`.

Reply:
Or you mean `invokeNoReturn`?

Reply:
It is; to get the return value of an `Invoke`, you have to call `getReturnValue()` on it.  The reason it is like this is that it is the only terminator which has a value.

Reply:
I'll handle `Invoke` once https://github.com/qbicc/qbicc/pull/1233 has been fixed.

Reply:
Sorry for delay, I'm adding `Invoke`...